app/assets/javascripts/darkswarm/directives/open_street_map.js.coffee

```diff
@@ -61,28 +61,47 @@ Darkswarm.directive 'ofnOpenStreetMap', ($window, Enterprises, EnterpriseModal,
 
     displayMap = ->
       setMapDimensions()
-      averageLatitude = averageAngle("latitude")
-      averageLongitude = averageAngle("longitude")
       zoomLevel = 6
       map = L.map('open-street-map')
       L.tileLayer.provider(openStreetMapProviderName, openStreetMapProviderOptions).addTo(map)
-      map.setView([averageLatitude, averageLongitude], zoomLevel)
+      map.setView([initialLatitude(), initialLongitude()], zoomLevel)
 
     displayEnterprises = ->
-      for enterprise in Enterprises.enterprises
-        if enterprise.latitude? && enterprise.longitude?
-          marker = buildMarker(enterprise, { lat: enterprise.latitude, lng: enterprise.longitude }, enterprise.name).addTo(map)
-          enterpriseNames.push(enterpriseName(enterprise))
-          markers.push(marker)
+      for enterprise in geocodedEnterprises()
+        marker = buildMarker(enterprise, { lat: enterprise.latitude, lng: enterprise.longitude }, enterprise.name).addTo(map)
+        enterpriseNames.push(enterpriseName(enterprise))
+        markers.push(marker)
+
+    disableSearchField = () =>
+      $('#open-street-map--search input').prop("disabled", true)
 
     displaySearchField = () ->
       new Autocomplete('#open-street-map--search',
         onSubmit: goToEnterprise
         search: searchEnterprises
       )
-      overwriteInlinePositionRelativeToPositionSearchField = ->
-        $('#open-street-map--search').css("position", "absolute")
-      overwriteInlinePositionRelativeToPositionSearchField()
+      overwriteInlinePositionRelativeToAbsoluteOnSearchField()
+      if geocodedEnterprises().length == 0
+        disableSearchField()
+
+    geocodedEnterprises = () ->
+      Enterprises.enterprises.filter (enterprise) ->
+        enterprise.latitude? && enterprise.longitude?
+
+    initialLatitude = () ->
+      if geocodedEnterprises().length > 0
+        averageAngle("latitude")
+      else
+        openStreetMapConfig.open_street_map_default_latitude || -37.4713077
+
+    initialLongitude = () ->
+      if geocodedEnterprises().length > 0
+        averageAngle("longitude")
+      else
+        openStreetMapConfig.open_street_map_default_longitude || 144.7851531
```

Comment:
imo this directive is already too long with too many responsibilities. Could you please extract some of these methods to a service?
can be under app/assets/javascripts/darkswarm/services/
You can extract averageAngle method or maybe the full displayMap method.

If you extract a service you will be able to add some unit tests that are missing.

Reply:
Another smaller thing, re the number here

144.7851531 is a magic number should be assigned to a variable with a name that explains what it is. Something like this?
defaultLongitude = 144.7851531

do we really need the value here? it is already set as the default on the server side.

Reply:
No problem @luisramos0 I extracted out a `MapCentreCalculator` service and added tests which is good because I think the averageAngle method was not working in certain scenarios. Although I haven't done a displayMap service yet.

I also removed the hardcoded latitude and longitude e.g 144.78. You're right it's probably unlikely that we need it.

Thanks

Reply:
oh, brilliant, that's it, thanks a lot!!
It's the best thing when we decide to write a few tests and end up finding a bug :muscle: